include/ipc_service/rpmsg_service.h

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_IPC_SERVICE_RPMSG_SERVICE_H_
+#define ZEPHYR_INCLUDE_IPC_SERVICE_RPMSG_SERVICE_H_
+
+#include <openamp/open_amp.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief RPMsg service API
+ * @defgroup rpmsg_service_api RPMsg service APIs
+ * @{
+ */
+
+/**
+ * @brief Register IPC endpoint
+ *
+ * Registers IPC endpoint to enable communication with a remote device.
+ * The endpoint is created when the slave device registers it.
+ *
+ * The same function registers endpoints for both master and slave devices.
+ *
+ * @param name String containing the name of the endpoint. Must be identical
+ *             for master and slave
+ * @param cb Callback executed when data are available on given endpoint
+ *
+ * @retval >=0 id of registered endpoint on success;
+ * @retval -ENOMEM when there is not enough slots to register the endpoint;
+ * @retval <0 an other negative errno code, reported by rpmsg.
+ */
+int rpmsg_service_register_endpoint(const char *name, rpmsg_ept_cb cb);
```

Comment:
To get rid of this dependency, you can define the callback structure as part of this API.

My suggestion:

```
enum ipc_service_endpoint_event_type {
	IPC_SERVICE_EVENT_CONNECTED,  // Handshake was successful
	IPC_SERVICE_EVENT_DATA,       // New packet arrived
};

struct ipc_service_endpoint_event {
	enum ipc_service_endpoint_event_type evt_type;
	union {
		struct bt_gatt_ots_oacp_create_params {
                        const uint8_t *buf;
                        size_t length;
		} data_params;
        }
};

typedef void (*ipc_service_endpoint_cb)(void *endpoint_context,
	struct ipc_service_endpoint_event *event, ...);

Reply:
I want to keep this PR simple. It is already much more complicated than I anticipated. Feel free to propose your changes in a separated PR after this one is merged